### PR TITLE
Contrastive Learning & Prior Latent Initialization

### PR DESCRIPTION
**Pull request by kohjiaxuan**
Added new code for contrastive learning & prior latent initialization
Docstrings available at front of each py file for explanation

**CONTRASTIVE LEARINING**
1. **Key python file: train_model_contrastive.py**
- Modified from train_model.py
- Besides the current reconstruction and regularization losses, train the model with contrastive loss. 
- Average species latent is the average latent for all latents under one species name
- For each epoch, pull all latents from each species closer to its average species latent, and push all latents away from other average species latent
- Modified to log the losses for each component per epoch and the % contribution for each loss
2. **Key CSV file: vtk_name_to_mapping_v2.csv**
- Contains all vtk filenames and their normalized position score (from 0 to 1) for model training and inference
3. vtk_parsing_logic.py
- Supporting code for train_model_contrastive.py, not called directly
- Regex to parse vtk filename to family, genus and species. Used by both training and classification code
4. train/train_deep_sdf_contrastive.py
- Supporting code for train_model_contrastive.py, not called directly
- Modified from train/train_deep_sdf.py
- Modified epoch function and loop to add contrastive loss calculation and backpropagation
5. datasets/sdf_dataset_contrastive.py
- Supporting code for train_model_contrastive.py, not called directly
- Modified from datasets/sdf_dataset.py
- Pass in the variable RUN_NAME to the cache path from train_model_contrastive RUN_NAME argument
- Ensures each run gets its own cache allowing for parallel training
- Remove hassle of manually modifying it every run

**PRIOR LATENT INITIALIZATION**
1. **Key python file: create_latents_per_group.py**
- Create average latent per family, genus or species and save as .npz file for classify_vertebrae_normpos.py to use
2. **Key python file: classify_vertebrae_normpos.py**
- Modified from classify_vertebrae.py
- Classify all test vertebrae and obtain results by pointing to a model run folder (passing argument) and its config file
- Allows for passing arguments to toggle between PCA (default), family, genus or species prior latent initialization
- Uses the new metric of calculation positional difference between test vertebrae and predicted Top 1 vertebrae. For example, if test vertebra is at 25% of spine (0.25) and predicted is at 30% of spine, error is |0.25-0.30|=0.05 (5%). Misclassified species will have 100% error (1.00). Average of 428 test vtk is taken for final positional error score
- Allows for passing arguments to toggle Gaussian noise for latent initialization or sampled xyz points for ablation studies (study robustness to noise)
3. **Key python file: aggregate_run_results.py**
- Run this after classify_vertebrae_normpos.py
- Pass folder path in RUN_NAME/classify_vertebrae/results in order to aggregate all results into a single CSV for metrics calculation, saved in RUN_NAME/classify_vertebrae/combined_results

### CODE REVIEW

File: NSM/datasets/sdf_dataset_contrastive.py
Comment (aubricot) [inline NSM/datasets/sdf_dataset_contrastive.py:1]: keep original sdf_dataset.py; changes only for handling of cache and time

File: train_model_contrastive.py
Comment (aubricot) [inline train_model_contrastive.py:1]: keep and incorporate changes into train_model.py; added argparse and conditional handling of NSM.train.train_deep_sdf_contrastive

**Review by aubricot — APPROVED** [review]
merge pull request, then delete sdf_dataset_contrastive.py and train_model_contrastive.py; update changes to train_model.py
